pinc/upload_file.inc

```diff
@@ -0,0 +1,276 @@
+    }
+    else
+    {
+        // resumable upload
+        $root_staging_dir = "/tmp/resumable_uploads";
+        $original_name = array_get($_POST, "resumable_filename", "");
+        $identifier = array_get($_POST, "resumable_identifier", "");
+        $hashed_filename = md5($identifier);
+        $file_path = "$root_staging_dir/$hashed_filename";
+        $file_size = filesize($file_path);
+        $file_info = ["name" => $original_name, "tmp_name" => $file_path];
+    }
+    // do some checks.
+    empty_check($file_size);
+    virus_check($file_path);
+    return $file_info;
+}
+
+function empty_check($file_size)
+{
+    if (0 == $file_size) {
+        throw new FileUploadException(_("File is empty"));
+    }
+}
+
+function zip_check($original_name, $file_path)
+{
+    if (substr($original_name, -4) != ".zip")
+    {
+        throw new FileUploadException( _("File name extension must be '.zip'") );
+    }
+    // ensure that it's a valid zip
+    // The extension was already checked and the file is not properly named (it has some temporary name), so we should
+    // disable the extension check.
+    if (!is_valid_zip_file($file_path, true))
+    {
+        throw new FileUploadException(sprintf(_("%s is not a valid zip file"),  $original_name));
+    }
+}
+
+function virus_check($file_path)
+{
+    global $antivirus_executable;
+    if(!$antivirus_executable)
+    {
+        return;
+    }
+    // perform '$antivirus_executable -- <FILENAME>' and expect return
+    // value = 0. we use -- to not parse any further arguments starting
+    // with -/-- as options
+    $av_test_result = array();
+    $av_retval=0;
+
+    $cmd = "$antivirus_executable -- " . escapeshellcmd($file_path);
+    exec($cmd, $av_test_result, $av_retval);
+    // $av_retval == 0 is ok
+    if ($av_retval == 1)
+    {
+        // Log the infected upload so that we can track user/frequency
+        $reporting_string = "Infected upload: " . $av_test_result[0];
+        error_log($reporting_string);
+        throw new FileUploadException(sprintf(_("AntiVirus: The scan reported an infection: %s. The upload has been discarded.
+            You should perform a complete virus scan on your computer as soon as possible."), $av_test_result[0]));
+    }
+    else if($av_retval > 1)
+    {
+        throw new FileUploadException(_("AntiVirus error:") . " " . $av_retval . "<br>" . $av_test_result[0]);
+    }
+}
+
+function make_valid_filename($name)
+{
+    // create a valid filename re-using the extension provided
+    $path_parts = pathinfo($name);
+    // replace starting hyphen or anything except alphanumeric or - or _ by _
+    $new_name = preg_replace('/^-|[^a-zA-Z0-9_-]/', '_', $path_parts['filename']) . "." . $path_parts["extension"];
+    $message = sprintf(_('File was renamed to "%1$s" because "%2$s" is not a valid filename'), $new_name, $name);
+    return [$new_name, $message];
+}
+
+function is_valid_filename($filename, $restrict_extension=False)
```

Comment:
It feels like `is_valid_filename()` should use `make_valid_filename()` to ensure we don't have two places that define what a valid filename is. Right now the regexes are different -- there is no limit on the size of the filename in `make_valid_filename()` but there is in `is_valid_filename()`.

Basically
```
function is_valid_filename($filename, $restrict_extension=False)
{
    return $filename == make_valid_filename($filename, $restrict_extension);
}
```
With appropriate updates in `make_valid_filename()` to handle `$restrict_extension`.

Reply:
Perhaps it would be better to remove both of these functions and have a javascript function to validate the filename before uploading (or whatever). Then the user would have the opportunity to rename the file himself.

Reply:
No, sorry, that won't work as a replacement. We can't replace server-side validation with client-side validation that can be bypassed. We could add a supplementary JS-side check, but that would require duplicating the validation logic -- which is not onerous but yet another thing to keep updated in two places.